fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt

```diff
@@ -71,10 +79,41 @@ class TranslationsDialogFragment : BottomSheetDialogFragment() {
             ),
         )
 
+        // Signalling user intention to translate
+        browserStore.dispatch(TranslationsAction.TranslateExpectedAction(args.sessionId))
```

Comment:
I don't know if this is okay to dispatch an action from a fragment. In all the project this is happening from a "feature" or from a Middleware.

Reply:
I'm having trouble finding a good place to put this or alternatives. I need the `tabId` for setting the error, so I need a tab related action for the `id`. Unfortunately, `InitAction` doesn't have the ID and some of the things I tried seem fragile. Let me know what ideas you have!

Tested changing [this middleware action](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt#39) to [AddTabAction](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt#177), this works fine for first user session, but the action isn't re-triggered when restoring from an app close. I can probably use [RestoreAction](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt#230) in addition to `AddTabAction`, but I think that could be potentially a lot of calls if the user had a bunch of tabs. [SelectTabAction](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt#205) sort of works on restore, but the initial first tab won't have data because it was never selected.

`TranslateExpectedAction` triggers now before we get here based on the translation engine thinking we should translate. This action that is firing from this Fragment is more of a guard to make sure the data is coming on the case where the user accesses via the browser menu. 

**Maybe the action should be triggered from the browser menu controller [there](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarMenuController.kt#410)?** This is the case I'm guarding against. WDYT?

(I plan on `TranslateExpectedAction` correlating to showing our toolbar icon as well. (I asked in [Figma](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD?node-id=7246:207835&mode=design#677883739), if we want to show the toolbar if the user accessed via browser menu. If setup and toolbar need decoupled, then it will be just a new action either way.)

Reply:
Any suggestions for this dispatch or is it okay for the time being? I can file a follow-up to look at this deeper too.

Reply:
I can take a deeper look at this later today if you'd like, but this comment:

> Tested changing [this middleware action](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt#39) to [AddTabAction](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt#177), this works fine for first user session, but the action isn't re-triggered when restoring from an app close. I can probably use [RestoreAction](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt#230) in addition to AddTabAction, but I think that could be potentially a lot of calls if the user had a bunch of tabs. [SelectTabAction](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt#205) sort of works on restore, but the initial first tab won't have data because it was never selected.

has me wondering if we can react to several actions (`SelectTab`, `RestoreTab` would maybe cover it?) in the middleware. My current understanding of the workflow in the diff is that we would wait until a tab is active and the dialog is pulled up to start the work of determining which translations are available for the dropdown. Is that correct? Do we want to start that work earlier so we don't have to potentially wait for it when the dialog becomes visible?

Reply:
But overall I think it would also be fine as is and we could revisit in a later bug if that's easiest for now

Reply:
> has me wondering if we can react to several actions (SelectTab, RestoreTab would maybe cover it?) 

I think there is some combination that would work. My main concern is fragility or missing some obscure case where neither action occurs. Also, I kinda wonder if `RestoreTab` could call this on a bunch of tabs and have a noticeable impact. I would need to dig a little deeper to feel confident about getting the right combination.

> My current understanding of the workflow in the diff is that we would wait until a tab is active and the dialog is pulled up to start the work of determining which translations are available for the dropdown. Is that correct?

Right, that is the main impact as far as the patch goes. There are some bigger pieces that need to harmonize too.

More background:
Two categories of translations:
* `Expected` - we will show a translate icon on the toolbar
   * Engine signals `TranslateExpectedAction`, which will probably correlate to showing icon on toolbar in [bug 1875817](https://bugzilla.mozilla.org/show_bug.cgi?id=1875817). (Originates [from delegate](https://searchfox.org/mozilla-mobile/rev/99576d6e69eac8ee1bf6c8a654d5fa96a6f6bf84/firefox-android/android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/translate/GeckoTranslateSessionDelegate.kt#26-28) -> [dispatches here ](https://searchfox.org/mozilla-mobile/rev/99576d6e69eac8ee1bf6c8a654d5fa96a6f6bf84/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/EngineObserver.kt#473).)
* `Unexpected` - no translate icon on toolbar, user accessed via browser menu on a site in their preferred language or an unsupported language
   * Criteria to show icon on toolbar as well in this case will happen if the user fully performs a translation from that entry point

I'm thinking `TranslateExpectedAction` will have to become `TranslateFetchLanguagesAction` or `TranslateStateInit` or use one of those global actions based on new information in the [Figma](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD?node-id=7246:207835&mode=design#677883739).

I completely forgot I already filed a bug to research this further when I wrote the middleware with [bug 1873672](https://bugzilla.mozilla.org/show_bug.cgi?id=1873672), so maybe we can brainstorm some ideas here and execute over there.

>Do we want to start that work earlier so we don't have to potentially wait for it when the dialog becomes visible?

Ideally, would like to get the languages early; however it is not an especially long operation.  (IIRC, first time ever is engine -> GV -> network call, but then later calls will be engine -> GV -> cache.) If it is okay to dispatch initialization here, I think it could conceptually simplify things as well.

Tangent:
We'll need to do similar initialization for [page settings](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD/Full-Page-Translation?type=design&node-id=7202-99696&mode=design&t=w3GVOGCG9Ytv9Zmj-0) and global settings [1](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD/Full-Page-Translation?type=design&node-id=7231-206407&mode=design&t=w3GVOGCG9Ytv9Zmj-0), [2](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD/Full-Page-Translation?type=design&node-id=7231-206408&mode=design&t=w3GVOGCG9Ytv9Zmj-0), [3](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD/Full-Page-Translation?type=design&node-id=7231-206406&mode=design&t=w3GVOGCG9Ytv9Zmj-0) , so maybe need to consider those situations as well and think of a plan.

1. Should the main translate page, page settings, and some very deep global options be initialized together on the same action, even if users are progressively less likely to go into the deeper settings? Or separate actions that occur as that data is needed?

Reply:
I lost my progress on a comment here by clicking a link 😩 hopefully this captures it all.

> Also, I kinda wonder if RestoreTab could call this on a bunch of tabs and have a noticeable impact.

We can always filter down to the selected tab (according to BrowserState, not the RestoreAction) in the middleware and launch the `TranslationsAction.TranslateExpectedAction` accordingly. 

Though, I am also seeing that `EngineObserver::onTranslateExpected` dispatches this action. Are we just trying to account for the unexpected cases here?

My big concern here is that we end up in one of these scenarios:
1. UI (toolbar button, dropdown, etc) is shown but isn't interact-able because we haven't finished fetching data. 
2. UI is not shown and pops in once state is available for observations.

Both cases can be avoided by fetching data eagerly enough, so the risk of a race condition is highest if we wait until the dialog is shown. I am not seeing placeholder UI in the Figma for things like the dropdown either, so I imagine we have to have the data available pretty immediately after a patch load.

Reply:
😩 Ohh, no, sorry you lost your original message! One nice thing about when we go to Phabricator is it saves draft comments and eagerly saves.
 
> We can always filter down to the selected tab (according to BrowserState, not the RestoreAction) in the middleware and launch the TranslationsAction.TranslateExpectedAction accordingly.

Ohh, that could be interesting and solve that issue!

> Are we just trying to account for the unexpected cases here?

Right, for the expected use-cases we should have already pulled the data almost right after page load. 

Since that is what I'm guarding, which is the browser menu use case, I wonder if I could launch the action [here](https://searchfox.org/mozilla-mobile/rev/07a850b0ef36fd7cd7aa339f6b14f40a91b6b4b2/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarMenuController.kt#410). Not super early, but maybe enough time?

Something I hadn't thought of is, I wonder if I could do a check if this value is empty and do it `onTranslationStateChange`. (It occurs fairly often, but I could check if the stored value is empty and fetch as needed. The languages stay the same between navigation as well.)

Thanks for the scenarios to watch for too and all the brainstorming ideas! I'll give this the attention it deserves in [bug 1873672](https://bugzilla.mozilla.org/show_bug.cgi?id=1873672). I like the idea of `AddTabAction` + filtered to selected tab `RestoreTab`.

Reply:
> Since that is what I'm guarding, which is the browser menu use case, I wonder if I could launch the action here. Not super early, but maybe enough time?

Conceptually I think this spot makes more sense, but I'm not sure we'd get measurably different results. I think it would work to add your dispatch here for now. From a long-term architectural viewpoint, I imagine we would eventually want this to look like

```
ToolbarMenu.Item.Translate -> {
    // now this action can be re-used by the translation middleware _and_ a navigation middleware
    store.dispatch(ToolbarMenuAction.TranslateClicked)
}
```

> Something I hadn't thought of is, I wonder if I could do a check if this value is empty and do it onTranslationStateChange.

Which value do you mean? Not totally following here

Reply:
Nice, I think a `TranslateClicked` would work well! Thanks for the idea.

> Which value do you mean? Not totally following here

I totally meant to remove that thought in an edit! Wasn't thinking clearly first thing in the morning. 😅